frontend/src/utils/d3/graph.js

```diff
@@ -0,0 +1,224 @@
+ * @param {int} y y of point outside circle
+ * @param {int} r radius of circle
+ * @returns x of point on circumference of circle
+ */
+function getXonCircumference(cx, cy, x, y, r) {
+  const angle = getAngle(cx, cy, x, y);
+  // cos takes radians
+  const adjusted_x = cx + Math.cos(angle) * r;
+  return adjusted_x;
+}
+
+/**
+ * Get the Y position of a point on the circumference of a circle
+ * @param {int} cx x of center of circle
+ * @param {int} cy y of center of circle
+ * @param {int} x x of point outside circle
+ * @param {int} y y of point outside circle
+ * @param {int} r radius of circle
+ * @returns y of point on circumference of circle
+ */
+function getYonCircumference(cx, cy, x, y, r) {
+  const angle = getAngle(cx, cy, x, y);
+  // sin takes radians
+  const adjusted_y = cy + Math.sin(angle) * r;
+  return adjusted_y;
+}
+
+/**
+ * Calculate the x and y of both edge ends as well as the quadratic curve point
+ * to make the edge curve
+ * @param {int} sourceX x of source node
+ * @param {int} sourceY y of source node
+ * @param {int} targetX x of target node
+ * @param {int} targetY y of target node
+ * @param {int} numEdges number of total edges
+ * @param {int} index index of edge to find its curve position
+ * @param {int} nodeRadius node radius
+ * @returns {obj} all the necessary points to make a curvy edge
+ */
+function getCurvedEdgePos(sourceX, sourceY, targetX, targetY, numEdges, index, nodeRadius) {
+  const arcWidth = Math.PI / 3;
+  const edgeStep = arcWidth / 5;
+  // get angle between nodes
+  const theta = getAngle(sourceX, sourceY, targetX, targetY);
+  // get adjusted angle based on edge index
+  const arc_p1 = theta + (edgeStep * index);
+  const arc_p2 = theta + Math.PI - (edgeStep * index);
+  // compute x's and y's
+  const x1 = sourceX + Math.cos(arc_p1) * nodeRadius;
+  const y1 = sourceY + Math.sin(arc_p1) * nodeRadius;
+  const x2 = targetX + Math.cos(arc_p2) * nodeRadius;
+  const y2 = targetY + Math.sin(arc_p2) * nodeRadius;
+  const alpha = 50; // tuning param
+  let l = (index * 0.1) + (index * 0.3); // num of edge
+  // move first arced edges out just a smidge
+  if (index === 1 || index === -1) {
+    l += (index * 0.4);
+  }
+  const bx = ((x1 + x2) / 2);
+  const by = (y1 + y2) / 2;
+  const vx = x2 - x1;
+  const vy = y2 - y1;
+
+  const norm_v = Math.sqrt(vx ** 2 + vy ** 2);
+  const vx_norm = vx / norm_v;
+  const vy_norm = vy / norm_v;
+  const vx_perp = -vy_norm;
+  const vy_perp = vx_norm;
+  const qx = bx + alpha * l * vx_perp;
+  const qy = by + alpha * l * vy_perp;
+  return {
+    x1, y1, qx, qy, x2, y2,
+  };
+}
+
+/**
+ * Get x and y positions on edge of node
+ * @param {int} cx connected node x position
+ * @param {int} cy connected node y position
+ * @param {int} x attached node x position
+ * @param {int} y attached node y position
+ * @param {int} r node radius
+ * @returns {obj} x and y postions on circumference of node
+ */
+function getXYonCircumference(cx, cy, x, y, r) {
```

Comment:
```suggestion
function getXYOnCircumference(cx, cy, x, y, r) {
```
So we have the combined method, but it's a bit wasteful since it computes the line angle twice.

Reply:
Updated in a224a58cecd993df9788e94249d0122dfa40abdc

Reply:
Latest changes here: 583354e471cf9c5c06b17d4059d8742dbbc3739c